eclipse-theia/theia
# Localize plugins using language packs

#### What it does

Addresses one of the open issues of #9538: Enables Theia to localize vscode extensions/plugins using language packs.

Basically addresses two problems regarding vscode extensions:

First off, it changes how the plugin package is read. Placeholders are now replaced when the frontend loads and not when the plugin is deployed. 

Secondly, through reverse engineering of the [`vscode-nls`](https://github.com/microsoft/vscode-nls/blob/main/src/node/main.ts) module, we are able to create `translationConfig` files that point to the correct localization file for each vscode extension. These files are cached in the `localization-cache` folder in the Theia home directory.

#### How to test

1. Install the `vscode.npm` builtin extension and a language pack of your liking.
2. Verify that the `npm Scripts` view is visible and in english.
3. Open a `package.json` file that contains a `scripts` section. Hover over one of the scripts and observe the `Run Script` popup: 
![image](https://user-images.githubusercontent.com/4377073/133004825-0d4a7ce1-8381-420f-a3ca-a4cef949dc87.png)
4. Use the `Configure Display Language` command to change to a different language.
5. The application reloads and the `npm Scripts` view should now have a different name. (*Tests that the `package.json` is still correctly translated*)
6. Hover over one of the scripts from the `package.json` file and observe a changed popup. (*Tests that the plugin host runtime is using the correct localization*) 
![image](https://user-images.githubusercontent.com/4377073/133004838-00c061f8-cc45-4824-855f-5395b5abd3bc.png)


#### Review checklist


#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
